swift/Arrow/Sources/Arrow/ArrowCImporter.swift

```diff
@@ -0,0 +1,172 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+import ArrowC
+
+public class ImportArrayHolder: ArrowArrayHolder {
+    let cArrayPtr: UnsafePointer<ArrowC.ArrowArray>
+    public var type: ArrowType {self.holder.type}
+    public var length: UInt {self.holder.length}
+    public var nullCount: UInt {self.holder.nullCount}
+    public var array: Any {self.holder.array}
+    public var getBufferData: () -> [Data] {self.holder.getBufferData}
+    public var getBufferDataSizes: () -> [Int] {self.holder.getBufferDataSizes}
+    public var getArrowColumn: (ArrowField, [ArrowArrayHolder]) throws -> ArrowColumn {self.holder.getArrowColumn}
+    private let holder: ArrowArrayHolder
+    init(_ holder: ArrowArrayHolder, cArrayPtr: UnsafePointer<ArrowC.ArrowArray>) {
+        self.cArrayPtr = cArrayPtr
+        self.holder = holder
+    }
+
+    deinit {
+        if self.cArrayPtr.pointee.release != nil {
+            ArrowCImporter.release(self.cArrayPtr)
+        }
+    }
+}
+
+public class ArrowCImporter {
+    private func appendToBuffer(
+        _ cBuffer: UnsafeRawPointer?,
+        arrowBuffers: inout [ArrowBuffer],
+        length: UInt) {
+        if cBuffer == nil {
+            arrowBuffers.append(ArrowBuffer.createEmptyBuffer())
+            return
+        }
+
+        let pointer = UnsafeMutableRawPointer(mutating: cBuffer)!
+        arrowBuffers.append(
+            ArrowBuffer(length: length, capacity: length, rawPointer: pointer, isMemoryOwner: false))
+    }
+
+    public init() {}
+
+    public func importType(_ cArrow: String, name: String = "") ->
+        Result<ArrowField, ArrowError> {
+        do {
+            let type = try ArrowType.fromCDataFormatId(cArrow)
+            return .success(ArrowField(name, type: ArrowType(type.info), isNullable: true))
+        } catch {
+            return .failure(.invalid("Error occurred while attempting to import type: \(error)"))
+        }
+    }
+
+    public func importField(_ cSchema: ArrowC.ArrowSchema) ->
+        Result<ArrowField, ArrowError> {
+        if cSchema.n_children > 0 {
+            return .failure(.invalid("Children currently not supported"))
+        } else if cSchema.dictionary != nil {
+            return .failure(.invalid("Dictinoary types currently not supported"))
+        }
+
+        switch importType(
+            String(cString: cSchema.format), name: String(cString: cSchema.name)) {
+        case .success(let field):
+            ArrowCImporter.release(cSchema)
+            return .success(field)
+        case .failure(let err):
+            ArrowCImporter.release(cSchema)
```

Comment:
If we're releasing on failure here, we should also release in the failure cases above (children and dictionary)?